Pass-through arguments to docker run and support --volume/-v …

meyerj opened the pull request:
… for the mount extension with backwards compatibility.

The overall goal is that the rocker command-line usage is closer to that of docker run and most arguments just work. Actually that makes some extensions like to pass environment variables obsolete.

However, it is not straight-forward because rocker cannot know which flags have one or more optional arguments and which positional argument is the image. That is why one has to split rocker and docker arguments by `--`, and ideally also the image name and optional command by another `--` argument:
```sh
$ rocker --help
usage: rocker [rocker arguments] [-- [docker run arguments]] -- image [command [command ...]]

A tool for running docker with extra options

positional arguments:
  image
  command

optional arguments:
...
```

Not sure whether there is a way to tell [argparse](https://docs.python.org/3/library/argparse.html) to stop searching for `-*` arguments after the first positional arguments, and consider all the rest as part of the command. At the moment this does not work, with or without this patch:
```sh
$ rocker ubuntu:focal bash -x
usage: rocker [rocker arguments] [-- [docker run arguments]] -- image [command [command ...]]
rocker: error: unrecognized arguments: -x
```

But [argparse does support the pseudo-argument `--`](https://docs.python.org/3/library/argparse.html#arguments-containing) which tells `parse_args()` that everything after that is a positional argument. So this works:
```sh
$ rocker -- ubuntu:focal bash -x
```

New version pushed to our internal repos as version `1:0.2.3-2intermodalics`.

meyerj committed:
Pass-through unknown arguments to docker run and support --volume/-v for the mount extension with backwards compatibility

The overall goal is that the rocker command-line usage is closer to that of docker run and most arguments just work.
However, it is not easily possible because rocker cannot know which flags have one or more optional arguments
and which positional argument is the image.

meyerj committed:
Pass-through unknown arguments to docker run and support --volume/-v for the mount extension with backwards compatibility

The overall goal is that the rocker command-line usage is closer to that of docker run and most arguments just work.
However, it is not easily possible because rocker cannot know which flags have one or more optional arguments
and which positional argument is the image.

meyerj committed:
Split command line at -- to separate rocker arguments from docker run arguments and positional arguments

meyerj committed:
Split command line at -- to separate rocker arguments from docker run arguments and positional arguments

meyerj committed:
v0.2.3-2intermodalics

fevb commented:
I get an error when trying to run the IM-VPS docker container:

```bash
Traceback (most recent call last):
  File "/usr/bin/rocker", line 11, in <module>
    load_entry_point('rocker==0.2.3', 'console_scripts', 'rocker')()
  File "/usr/lib/python3/dist-packages/rocker/cli.py", line 95, in main
    return dig.run(**args_dict, additional_args=docker_run_args)
  File "/usr/lib/python3/dist-packages/rocker/core.py", line 256, in run
    docker_args += e.get_docker_args(self.cliargs)
  File "/usr/lib/python3/dist-packages/rocker/mount_extension.py", line 75, in get_docker_args
    elems = volume.split(':')
AttributeError: 'list' object has no attribute 'split'
```

Unfortunately I cannot revert back to version `1:0.2.3-1intermodalics`:

```bash
fevb@dell-home:~/im-vps$ sudo apt install python3-rocker=1:0.2.3-2intermodalics
Reading package lists... Done
Building dependency tree       
Reading state information... Done
python3-rocker is already the newest version (1:0.2.3-2intermodalics).
0 upgraded, 0 newly installed, 0 to remove and 2 not upgraded.
fevb@dell-home:~/im-vps$ sudo apt install python3-rocker=1:0.2.3-1intermodalics
Reading package lists... Done
Building dependency tree       
Reading state information... Done
E: Version '1:0.2.3-1intermodalics' for 'python3-rocker' was not found
```

meyerj committed:
Fix parsing --mount arguments with : for backwards compatibility

Fixes:

Traceback (most recent call last):
  File "/usr/bin/rocker", line 11, in <module>
    load_entry_point('rocker==0.2.3', 'console_scripts', 'rocker')()
  File "/usr/lib/python3/dist-packages/rocker/cli.py", line 95, in main
    return dig.run(**args_dict, additional_args=docker_run_args)
  File "/usr/lib/python3/dist-packages/rocker/core.py", line 256, in run
    docker_args += e.get_docker_args(self.cliargs)
  File "/usr/lib/python3/dist-packages/rocker/mount_extension.py", line 75, in get_docker_args
    elems = volume.split(':')
AttributeError: 'list' object has no attribute 'split'

meyerj committed:
v0.2.3-3intermodalics

meyerj commented:
> I get an error when trying to run the IM-VPS docker container:

Should be fixed in https://github.com/Intermodalics/rocker/pull/5/commits/9a34253983b8ba48fc0d42acda39c1ac8a278bc7. New debian version pushed.

meyerj committed:
Merge remote-tracking branch 'meyerj/feature/argument-pass-through' into feature/argument-pass-through

meyerj committed:
Merge branch 'master' into feature/argument-pass-through

meyerj committed:
Fix closing parenthesis in generate_docker_cmd()

meyerj committed:
Merge remote-tracking branch 'meyerj/feature/argument-pass-through' into feature/argument-pass-through

meyerj committed:
Remove short option -v for --version which is conflicting with the mount extension

meyerj committed:
Add extra space for additional_docker_args in generate_docker_cmd()

meyerj committed:
Merge remote-tracking branch 'meyerj/feature/argument-pass-through' into feature/argument-pass-through

meyerj committed:
v0.2.3-4intermodalics
